### PR TITLE
core: record stats for GET requests

### LINKED ISSUE #2896: AbstractClientStream2.GetFramer doesn't use statsTraceCtx
GetFramer doesn't use the normal MessageFramer, so it should probably call statsTraceCtx and report the request message size and similar.

### PR DESCRIPTION
**Pull request by zhangkun83**
Also define the expected call sites of StatsTraceContext.

Resolves #2896

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 1]
```diff
@@ -354,6 +354,18 @@ public void writePayload(InputStream message) {
...
+        statsTraceCtx.outboundMessage();
+        statsTraceCtx.outboundUncompressedSize(payload.length);
+        // NB(zhangkun83): this is not accurate, because the underlying transport will probably
+        // encode it using e.g., base64.  However, we are not supposed to know such detail here.
+        //
+        // We don't want to move this line to where the encoding happens either, because we'd better
+        // contain the message stats reporting in Framer as suggested in StatsTraceContext.
+        // Scattering the reporting sites increases the risk of mis-counting or double-counting.
+        //
+        // Because the payload is usually very small, and people shouldn't care about the size
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: s/and//

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 2]
```diff
@@ -354,6 +354,18 @@ public void writePayload(InputStream message) {
       Preconditions.checkState(payload == null, "writePayload should not be called multiple times");
       try {
         payload = IoUtils.toByteArray(message);
+        statsTraceCtx.outboundMessage();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Move after the try-catch?

File: core/src/main/java/io/grpc/internal/StatsTraceContext.java [hunk 3]
```diff
@@ -104,6 +104,8 @@ public static StatsTraceContext newServerContext(
 
   /**
    * See {@link ClientStreamTracer#headersSent}.  For client-side only.
+   *
+   * <p>Transport-specific thus should be called by transport implementations.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/StatsTraceContext.java]: Add a comma before thus?

**Comment by zhangkun83** [comment]
Thanks! All suggestions applied.

**Pull request merged**